### ISSUE TITLE
Feature Request: Upgrade ISO-8

### ISSUE THREAD

**Issue by dwiza3962** [tmarenko/mff_auto#42]
Similar to Upgrade Gear, would be nice to auto upgrade iso-8 sets as inventory hoarding is terrible.

Would like ability to upgrade each type to say 4*/5* or so.
Would like ability to select which types can be used for upgrade.
Number of Times to upgrade and upgrade all.

**Comment by akshaypune** [comment]
It would be really helpful to have ability to enhance and even combine ISOs in order to finish Daily quests and special events which need these. Ability to upgrade only 4/5/6* ISOs would also do.

**Comment by tmarenko** [comment]
I came up with idea how to do upgrades and combines. Basically with limited tools of text/image recognition there is no easy way to determine ISO-8 rank.
The logic of automatization would be next:
![image](https://user-images.githubusercontent.com/31345208/134290887-4eb98026-cd98-4b41-88f3-cb2d94378042.png)
From bottom right to top left bot will be clicking on each of ISO-8 and will be checking whether this ISO-8 is available for upgrading/combining.
You could choose which types you want to upgrade/combine and which types/which stars you would waste to do that:
![image](https://user-images.githubusercontent.com/31345208/134291008-3ebd565f-12bd-4394-8bc6-597c00398dca.png)
The ISO-8 for combining would be always the top one:
![image](https://user-images.githubusercontent.com/31345208/134291103-e71fbd65-faca-4e6c-b630-1995a3acd73d.png)

This method is **not optimal** but is easy one to automate.
In short, you could choose:

- types of ISO-8 to perform upgrade/combine (white ISO, red ISO, etc.)
- types and rank of ISO-8 to waste into upgrading
- how many times to upgrade/combine for each of selected type

Unable to choose:
- rank of upgraded ISO-8 (it would be rank-best available ISO-8 selected from your inventory)
- rank of combining material (it would be the first one available)

What do you think about this? Is this okay for your request or may be you have suggestion?

**Comment by dwiza3962** [comment]
Thank you for looking into this.    I think this would work.   I didn't think how hard it would be to determine rank - but since it allows sorting by rank - I agree with your approach the first one (bottom right) being best available.   

The only suggestion I'm thinking of that might get more complicated is to use text recognition on the ISO-8 and provide options to lock certain conditions - Also understanding how it could handle a locked scenario.

A couple use cases.
UC1: I tend to lock/combine all my chaotics that have any of the attack increases and HP Increase or 'Physical Attack Increase'
'Physical Attack Increase' && 'HP Up'
'Energy Attack Increase' && 'HP Up'
'All Basic Attack Increase' && 'HP Up'
No matter the rank since these can't be rerolled.

UC2: 
I tend to want to keep/upgrade
All Defense but Waste Energy Defense (Blues)
All Defense but Waste Physical Defense (Greens)
No matter the rank.

Not sure if that specific stats could be an option in the UI to automatically lock or waste.   Also how would this process work if certain ISO-8's are locked?  Is there a way to detect if something is locked?  Anything is better than nothing at this point.    I'm trying to keep building my account and focusing on content rather than inventory management.

**Comment by tmarenko** [comment]
> I tend to want to keep/upgrade
> All Defense but Waste Energy Defense (Blues)
> All Defense but Waste Physical Defense (Greens)
> 
> Not sure if that specific stats could be an option in the UI to automatically lock or waste

I think the best way is to use 'Quick Enhance' menu for upgrading. This would allow to use any selected type/rank to waste but this menu doesn't have info about ISO's options:
![image](https://user-images.githubusercontent.com/31345208/134467805-f884f9ab-c1ac-4d74-af17-b70b94ade52a.png)
This menu skips locked ISOs so that wouldn't be a problem.

> Also how would this process work if certain ISO-8's are locked?

For combining there is notification when you try to use locked ISO as combining material:
![image](https://user-images.githubusercontent.com/31345208/134467970-d604e00c-dfe5-4aa9-93da-4e6e3f28a7ce.png)
In case it happens bot would skip current ISO and try to combine next one in the inventory.

> I tend to lock/combine all my chaotics that have any of the attack increases and HP Increase or 'Physical Attack Increase'

That probably is doable as another action/feature (before upgrading/combining).
Same logic as upgrading: click on each of ISO-8 from bottom right to top left, parse ISO's option, lock if option is desirable.

**Comment by tmarenko** [comment]
Added in https://github.com/tmarenko/mff_auto/commit/04d0eaff02948c51cec3025e25aaa8abe4eec8e9
Released in https://github.com/tmarenko/mff_auto/releases/tag/0.9.0

**Comment by dwiza3962** [comment]
Thank you! Will be trying this out today.
